### DISCUSSION TITLE
Config Parser Rework

### DISCUSSION THREAD

**Discussion by borgmanJeremy** [flameshot-org/flameshot#1862] [General]
This discussion is to address any issues found during the config parser rework

**Comment by veracioux** [comment]
I think I have fixed the issue. Please check.

**Reply by borgmanJeremy** [reply]
Yup that fixed it!, ill keep testing

**Comment by veracioux** [comment]
@mmahmoudian Replying to https://github.com/flameshot-org/flameshot/pull/1859#issuecomment-911806662

> I built from source against [b6641ea](https://github.com/flameshot-org/flameshot/commit/b6641eaf4a6a66426bfa40cd6df1a6ad1b3085a0) commit and:
> 
> * having `drawFontSize=A` didn't raise an error
> 
> * having `uploadHistoryMax=2.5` (with decimal points) would automagically correct it to `uploadHistoryMax=0` which should not become 0
> 
> * having`TYPE_RESIZE_RIGHT=Shift+Right:` (extra colon at the end) would not generate error, it is not automagically fixed and the shortcut does not work anymore. even the colon shows up in the config window:

I can't reproduce the first two, they work as expected on my machine:
- `drawFontSize=A` raises an error
- `uploadHistoryMax=2.5` raises an error and no correction happens

Keep in mind that the error notification is sent only when the file **changes** from no_error to error. If the file already had an error and you introduce a new one, no notification.

 The third one is still a TODO.

What exact steps did you take?
- Did the config contain the error before you started the daemon, or did you change the config while the daemon was running?
- If the latter, was the config window open when the file changed, or the capture GUI, or none?
- When does flameshot overwrite 2.5 with 0: immediately after the file changed or after you interact with the config window?
- Can you post your exact config before you introduced the error?

**Reply by mmahmoudian** [reply]
> Did the config contain the error before you started the daemon, or did you change the config while the daemon was running?

I changed the config after the daemon was started 

> Did the config contain the error before you started the daemon, or did you change the config while the daemon was running?

I had the config file open in neovim, so I just `:x` , then started capturing an area in gui mode, then cancelled pressing Esc, then opened the config in neovim again and it was turned to 0

> Can you post your exact config before you introduced the error?

I don't have access to that computer and you should wait until tomorrow. I'll update this post

**Reply by veracioux** [reply]
> > Did the config contain the error before you started the daemon, or did you change the config while the daemon was running?
> 
> I changed the config after the daemon was started
> 
> > Did the config contain the error before you started the daemon, or did you change the config while the daemon was running?
> 
> I had the config file open in neovim, so I just `:x` , then started capturing an area in gui mode, then cancelled pressing Esc, then opened the config in neovim again and it was turned to 0
> 
> > Can you post your exact config before you introduced the error?
> 
> I don't have access to that computer and you should wait until tomorrow. I'll update this post

Is this still an issue?

**Reply by mmahmoudian** [reply]
Well I posted my config in the other thread. I should check with the latest changes and report back.

**Comment by mmahmoudian** [comment]
@veracioux I built against 3d8e89b from  https://github.com/veracioux/flameshot and faced a funny situation. My current config that works just fine with v0.10.1 is now:
![image](https://user-images.githubusercontent.com/390889/132826273-46b64e23-91fe-40ce-bc6c-3d4551abc395.png)

This got me laughing so hard, because now I know how it really feels to have no clue what is wrong with my config 🤣  The config looks correct and I cannot spot any mistakes:

<details>
<summary>Content of my <code>~/.config/flameshot/flameshot.ini</code></summary>

```ini
[General]
buttons=@Variant(\0\0\0\x7f\0\0\0\vQList<int>\0\0\0\0\x14\0\0\0\0\0\0\0\x1\0\0\0\x2\0\0\0\x3\0\0\0\x4\0\0\0\x5\0\0\0\x6\0\0\0\x13\0\0\0\x10\0\0\0\f\0\0\0\x14\0\0\0\a\0\0\0\b\0\0\0\t\0\0\0\x11\0\0\0\n\0\0\0\v\0\0\0\xf\0\0\0\r\0\0\0\x12)
contrastOpacity=188
disabledTrayIcon=false
drawColor=#232629
drawFontSize=5
drawThickness=2
saveAfterCopyPath=/home/mehrad/Pictures
savePath=/home/mehrad/Pictures
savePathFixed=false
setSaveAsFileExtension=.png
showStartupLaunchMessage=true
startupLaunch=true
uiColor=#ff7300
undoLimit=100
uploadHistoryMax=0
useJpgForClipboard=false

[Shortcuts]
TYPE_ARROW=A
TYPE_CIRCLE=
TYPE_CIRCLECOUNT=
TYPE_COMMIT_CURRENT_TOOL=Ctrl+Return
TYPE_COPY=Ctrl+C
TYPE_DELETE_CURRENT_TOOL=Del
TYPE_DRAWER=D
TYPE_EXIT=Ctrl+Q
TYPE_IMAGEUPLOADER=
TYPE_MARKER=M
TYPE_MOVESELECTION=Ctrl+M
TYPE_MOVE_DOWN=Down
TYPE_MOVE_LEFT=Left
TYPE_MOVE_RIGHT=Right
TYPE_MOVE_UP=Up
TYPE_OPEN_APP=Ctrl+O
TYPE_PENCIL=
TYPE_PIN='
TYPE_PIXELATE=B
TYPE_RECTANGLE=R
TYPE_REDO=Ctrl+Shift+Z
TYPE_RESIZE_DOWN=Shift+Down
TYPE_RESIZE_LEFT=Shift+Left
TYPE_RESIZE_RIGHT=Shift+Right
TYPE_RESIZE_UP=Shift+Up
TYPE_SAVE=Ctrl+S
TYPE_SELECTION=S
TYPE_SELECTIONINDICATOR=
TYPE_SELECT_ALL=Ctrl+A
TYPE_TEXT=T
TYPE_TOGGLE_PANEL=Space
TYPE_UNDO=Ctrl+Z
```

</details>

Note that the `'` is intentional as single quote looks like pin and is also conveniently beside <kbd>Enter</kbd> in ISO keyboard layout.


I think at some point we should also tell user what is wrong with their config

**Reply by veracioux** [reply]
@mmahmoudian I found out where the problems are:
1. I had set that `uploadHistoryMax` must be >= 1. Even before my PR your value of zero would be saturated to 1, but it didn't report an error. I changed it so that it must be >= 0 now.
2. When I merged master I accidentally removed `TYPE_INVERT` entry in one place so the `buttons` entry in your config was considered erroneous.

I fixed the issues. You should not be getting an error anymore. But even after the fix, your button configuration could be messed up. I have made it so that this won't happen after #1880 gets merged into master. I have applied the same change in this PR. While you are testing the config checker PR, the issue may persist if you keep changing the flameshot version you use.
